utils.go

```diff
@@ -14,14 +43,208 @@ type Storage struct {
+		_, host, port = ep.HTTPS()
+		secure = true
+	default:
+		return nil, services.PairUnsupportedError{Pair: ps.WithEndpoint(opt.Endpoint)}
+	}
+	url := fmt.Sprintf("%s:%d", host, port)
+
+	srv.service, err = minio.New(url, &minio.Options{
+		Creds:  credentials.NewStaticV4(ak, sk, ""),
+		Secure: secure,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if opt.HasDefaultServicePairs {
+		srv.defaultPairs = opt.DefaultServicePairs
+	}
+	if opt.HasServiceFeatures {
+		srv.features = opt.ServiceFeatures
+	}
+
+	return
+}
+
+func newServicerAndStorager(pairs ...types.Pair) (srv *Service, store *Storage, err error) {
+	srv, err = newServicer(pairs...)
+	if err != nil {
+		return
+	}
+
+	store, err = srv.newStorage(pairs...)
+	if err != nil {
+		err = services.InitError{Op: "new_storager", Type: Type, Err: formatError(err), Pairs: pairs}
+		return nil, nil, err
+	}
+	return srv, store, nil
+}
+
+func formatError(err error) error {
+	if _, ok := err.(services.InternalError); ok {
+		return err
+	}
+
+	e, ok := err.(minio.ErrorResponse)
+	if ok {
+		switch e.Code {
+		case "AccessDenied":
+			return fmt.Errorf("%w, %v", services.ErrPermissionDenied, err)
+		case "NoSuchKey":
+			return fmt.Errorf("%w, %v", services.ErrObjectNotExist, err)
+		default:
+			return fmt.Errorf("%w, %v", services.ErrUnexpected, err)
+		}
+	}
+
+	return fmt.Errorf("%w, %v", services.ErrUnexpected, err)
+}
+
+func (s *Service) newStorage(pairs ...types.Pair) (st *Storage, err error) {
+	opt, err := parsePairStorageNew(pairs)
+	if err != nil {
+		return nil, err
+	}
+
+	store := &Storage{
+		client:  s.service,
+		bucket:  opt.Name,
+		workDir: "/",
+	}
+
+	if opt.HasWorkDir {
+		if !strings.HasSuffix(opt.WorkDir, "/") {
+			opt.WorkDir += "/"
+		}
+		store.workDir = opt.WorkDir
+	}
+	if opt.HasDefaultStoragePairs {
+		store.defaultPairs = opt.DefaultStoragePairs
+	}
+	if opt.HasStorageFeatures {
+		store.features = opt.StorageFeatures
+	}
+
+	return store, nil
+}
+
+func (s *Service) formatError(op string, err error, name string) error {
+	if err == nil {
+		return nil
+	}
+
+	return services.ServiceError{
+		Op:       op,
+		Err:      formatError(err),
+		Servicer: s,
+		Name:     name,
+	}
 }
 
 func (s *Storage) formatError(op string, err error, path ...string) error {
-	panic("implement me")
+	if err == nil {
+		return nil
+	}
+
+	return services.StorageError{
+		Op:       op,
+		Err:      formatError(err),
+		Storager: s,
+		Path:     path,
+	}
+}
+
+// getAbsPath will calculate object storage's abs path
+func (s *Storage) getAbsPath(path string) string {
+	prefix := strings.TrimPrefix(s.workDir, "/")
+	return prefix + path
+}
+
+// getRelPath will get object storage's rel path.
+func (s *Storage) getRelPath(path string) string {
+	prefix := strings.TrimPrefix(s.workDir, "/")
+	return strings.TrimPrefix(path, prefix)
+}
+
+func (s *Storage) formatFileObject(v minio.ObjectInfo) (o *types.Object, err error) {
+	if v.ETag == "" {
+		return nil, services.ErrObjectNotExist
```

Comment:
Like mentioned [before](https://github.com/beyondstorage/go-service-minio/pull/6/files#r668456350), why we check `v.Etag` here and return `ErrObjectNotExist` when etag is blank?

Reply:
ditto